### PR TITLE
Issues 2360 (#1)

### PR DESCRIPTION
**Pull request by LeetaoGoooo**
## Pull Request Checklist

- [ ] A news fragment is added in `news/` describing what is new.
- [ ] Test cases added for changed code.

## Describe what you have changed in this PR.

Fix bugs mentioned in https://github.com/pdm-project/pdm/issues/2360, do some check before init a project, follow [PEP 426](https://peps.python.org/pep-0426/#name) and sanitize the default name by replacing the illegal characters

### CODE REVIEW

File: src/pdm/cli/commands/init.py [hunk 1]
```diff
@@ -90,6 +90,21 @@ def ask(self, question: str, default: str) -> str:
             return default
         return termui.ask(question, default=default)
 
+    def ask_project(self, project: Project) -> str:
+        name = self.ask("Project name", project.root.name)
+        if validate_project_name(name):
+            return name
+        project.core.ui.echo(
+            "Project name is not validate \n"
```
Comment (pawamoy) [inline src/pdm/cli/commands/init.py:98]:
```suggestion
            "Project name is not valid \n"
```

File: src/pdm/utils.py [hunk 2]
```diff
@@ -453,3 +453,26 @@ def package_installed(package_name: str) -> bool:
         return False
     else:
         return True
+
+
+def validate_project_name(name: str) -> bool:
+    """Check if the project name is validate or not"""
```
Comment (pawamoy) [inline src/pdm/utils.py:459]:
```suggestion
    """Check if the project name is valid or not."""
```

File: src/pdm/utils.py [hunk 3]
```diff
@@ -453,3 +453,26 @@ def package_installed(package_name: str) -> bool:
...
+
+    allowed_chars = set("abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789_-")
```
Comment (pawamoy) [inline src/pdm/utils.py:461]: I suppose we could use `string.ascii_letters` and `string.digits` instead:

```suggestion
    allowed_chars = set(string.ascii_letters + string.digits + "_-")
```

Comment (LeetaoGoooo) [reply src/pdm/utils.py:461]: yeah, you're right, it's much better ~~ anything else

File: src/pdm/utils.py [hunk 4]
```diff
@@ -453,3 +453,27 @@ def package_installed(package_name: str) -> bool:
...
+
+    if keyword.iskeyword(name):
+        return False
+
+    import builtins
+
+    if name in dir(builtins):
+        return False
+
+    return True
```
Comment (frostming) [inline src/pdm/utils.py:479]: We can only consider the standard of [PEP 426](https://peps.python.org/pep-0426/#name) which even gives a regex pattern that you can use. No need to check the keywords and builtin.

File: src/pdm/cli/commands/init.py [hunk 5]
```diff
@@ -90,6 +90,21 @@ def ask(self, question: str, default: str) -> str:
             return default
         return termui.ask(question, default=default)
 
+    def ask_project(self, project: Project) -> str:
+        name = self.ask("Project name", project.root.name)
```
Comment (frostming) [inline src/pdm/cli/commands/init.py:94]: Also please don't validate the default value or user may feel confused.
Instead, we sanitize the default name by replacing the illegal characters.

**Comment by frostming** [comment]
Fine, can you please add a news fragment under `news` and we are good to go.

**Comment by LeetaoGoooo** [comment]
> Fine, can you please add a news fragment under `news` and we are good to go.

It is completed. Thank you for your patient guidance and assistance.～～

**Review by frostming — APPROVED** [review]

**Pull request merged**